File: python/nbody/parallel_simulation.py
import csv
import numpy as np
from typing import List, Optional
from multiprocessing import Pool, cpu_count
from .body import Body
from .io_utils import write_state

G = 1.0
EPSILON = 1e-2
MAX_SPEED = 50.0

def compute_forces_chunk(args):
    bodies, start, end = args

    local_forces = np.zeros((end - start, 2))

    for local_i, i in enumerate(range(start, end)):
        bi = bodies[i]
        force = np.zeros(2)

        for j, bj in enumerate(bodies):
            if i == j:
                continue

            diff = bj.position - bi.position
            dist_sq = np.dot(diff, diff) + EPSILON
            dist = np.sqrt(dist_sq)

            force_mag = G * bi.mass * bj.mass / dist_sq
            force_vec = force_mag * diff / dist

            force += force_vec

        local_forces[local_i] = force

    return start, local_forces

def compute_forces_parallel(bodies: List[Body], pool, num_workers: int) -> None:
    n = len(bodies)

    chunk_size = (n + num_workers - 1) // num_workers

    tasks = []
    for p in range(num_workers):
        start = p * chunk_size
        end = min((p + 1) * chunk_size, n)

        if start < end:
            tasks.append((bodies, start, end))

    results = pool.map(compute_forces_chunk, tasks)

    for start, local_forces in results:
        for offset, f in enumerate(local_forces):
            bodies[start + offset].force = f

def update_bodies(bodies: List[Body], dt: float) -> None:
    for b in bodies:
        acceleration = b.force / b.mass
        b.velocity += acceleration * dt

        speed = np.linalg.norm(b.velocity)
        if speed > MAX_SPEED:
            b.velocity *= MAX_SPEED / speed

        b.position += b.velocity * dt

def simulate_parallel(bodies: List[Body], steps: int, dt: float, num_workers: int, writer: Optional[csv.writer] = None) -> None:
    with Pool(cpu_count()) as pool:
        for step in range(steps):
            for b in bodies:
                b.reset_force()

            compute_forces_parallel(bodies, pool, num_workers)

            update_bodies(bodies, dt)

            if writer is not None:
                write_state(writer, step, bodies)

File: rust/src/parallel_simulation.rs
use std::fs::File;
use std::io::BufWriter;
use std::sync::Arc;
use std::thread;

use crate::body::Body;

pub const G: f64 = 1.0;
pub const EPSILON: f64 = 1e-5;

pub fn compute_forces_parallel(bodies: &mut [Body], num_threads: usize) {
    let n = bodies.len();

    // Reset forces
    for b in bodies.iter_mut() {
        b.reset_force();
    }

    // Shared read-only snapshot
    let bodies_arc = Arc::new(bodies.to_vec());

    let chunk_size = (n + num_threads - 1) / num_threads;

    let mut handles = vec![];

    for t in 0..num_threads {
        let bodies_clone = Arc::clone(&bodies_arc);

        let start = t * chunk_size;
        let end = ((t + 1) * chunk_size).min(n);

        let handle = thread::spawn(move || {
            // Each thread keeps FULL force array (important!)
            let mut local_forces = vec![[0.0; 2]; bodies_clone.len()];

            for i in start..end {
                for j in (i + 1)..bodies_clone.len() {
                    let dx = bodies_clone[j].position[0] - bodies_clone[i].position[0];
                    let dy = bodies_clone[j].position[1] - bodies_clone[i].position[1];

                    let dist_sq = dx * dx + dy * dy + EPSILON;
                    let dist = dist_sq.sqrt();

                    let force_mag =
                        G * bodies_clone[i].mass * bodies_clone[j].mass / dist_sq;

                    let fx = force_mag * dx / dist;
                    let fy = force_mag * dy / dist;

                    // Apply to both i and j (symmetry)
                    local_forces[i][0] += fx;
                    local_forces[i][1] += fy;

                    local_forces[j][0] -= fx;
                    local_forces[j][1] -= fy;
                }
            }

            local_forces
        });

        handles.push(handle);
    }

    // Merge all thread results
    let mut total_forces = vec![[0.0; 2]; n];

    for handle in handles {
        let local = handle.join().unwrap();

        for i in 0..n {
            total_forces[i][0] += local[i][0];
            total_forces[i][1] += local[i][1];
        }
    }

    // Write back to bodies
    for i in 0..n {
        bodies[i].force = total_forces[i];
    }
}

pub fn update_bodies(bodies: &mut [Body], dt: f64) {
    for b in bodies.iter_mut() {
        let ax = b.force[0] / b.mass;
        let ay = b.force[1] / b.mass;

        b.velocity[0] += ax * dt;
        b.velocity[1] += ay * dt;

        b.position[0] += b.velocity[0] * dt;
        b.position[1] += b.velocity[1] * dt;
    }
}

pub fn simulate_parallel(
    bodies: &mut [Body],
    steps: usize,
    dt: f64,
    num_threads: usize,
    mut writer: Option<&mut BufWriter<File>>,
) -> std::io::Result<()> {
    for step in 0..steps {
        compute_forces_parallel(bodies, num_threads);
        update_bodies(bodies, dt);

        if let Some(w) = writer.as_deref_mut() {
            crate::io_utils::write_state(w, step, bodies)?;
        }
    }
    Ok(())
}